api/models/datastore.go

```diff
@@ -5,21 +5,66 @@ import (
 	"errors"
 )
 
+//TODO path name validation: implementations currently assume/require/rely on:
+// - leading slash
+// - overall legal/scrubbed path (no relative paths, consecutive slashes, etc)
+// - trailing slash allowed and distinct from none
 type Datastore interface {
-	// GetApp returns the app called appName or nil if it doesn't exist
+
+	// GetApp gets an App by name.
+	// Returns ErrDatastoreEmptyAppName for empty appName.
+	// Returns ErrAppsNotFound if no app is found.
 	GetApp(ctx context.Context, appName string) (*App, error)
+
+	// GetApps gets a slice of Apps, optionally filtered by name.
+	// Missing filter or empty name will match all Apps.
 	GetApps(ctx context.Context, filter *AppFilter) ([]*App, error)
+
+	// InsertApp inserts an App. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsAlreadyExists if an App by the same name already exists.
 	InsertApp(ctx context.Context, app *App) (*App, error)
+
+	// UpdateApp updates an App's Config. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsNotFound if an App is not found.
+	//TODO are app.Config's fields added to the existing field or do they replace them?
 	UpdateApp(ctx context.Context, app *App) (*App, error)
```

Comment:
Patching should be doable. (The current postgres impl for this and UpdateRoute both currently act as 'setters', but shouldn't become too complicated to fix).

The distinction between nil and not found is regarding the ```app``` parameter vs. ```App``` the type - a nil parameter is an illegal argument, while trying to update an App by name that doesn't exist will result in that particular App not being found.  Maybe it's confusing for the doc to reference the param name as ```app``` would ``` `app` ``` make it more clear?

When you say ```like it used to```, which implementation are you referring too?  Most of these error docs are just documenting existing behavior, or trying to guess at which is correct between inconsistent implementations or method behavior (for example, the postgres impl already returned ErrAppsNotFound when no rows were updated).  Without a broader perspective on how this API fits into the project (or even if it's public?) I don't have strong opinions either way, and these TODOs are sort of the points where I couldn't align the impls without deciding one way or the other, so whatever you think is best.